### PR TITLE
Make DeployGrpcService.visualizeDag stack safe

### PR DESCRIPTION
**Pull request by sebtomba**
## Overview
Make DeployGrpcService.visualizeDag stack safe

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3236

### CODE REVIEW

File: graphz/src/main/scala/coop/rchain/graphz/Graphz.scala [hunk 1]
```diff
@@ -10,16 +10,14 @@ trait GraphSerializer[F[_]] {
...
-    MonadState[F, StringBuffer].modify(sb => sb.append(str + suffix))
+    Defer[F].defer(MonadState[F, StringBuffer].modify(sb => sb.append(str + suffix)))
 }
 
-class ListSerializer[F[_]: Applicative: MonadState[?[_], Vector[String]]]
-    extends GraphSerializer[F] {
+class ListSerializer[F[_]: Defer: MonadState[?[_], Vector[String]]] extends GraphSerializer[F] {
   def push(str: String, suffix: String): F[Unit] =
-    MonadState[F, Vector[String]].modify(_ :+ (str + suffix))
+    Defer[F].defer(MonadState[F, Vector[String]].modify(_ :+ (str + suffix)))
```
Comment (ArturGajowy) [inline graphz/src/main/scala/coop/rchain/graphz/Graphz.scala]: I don't think the `defer`-s are needed here. Unless you've seen this fail with Eval and without the `defer`-s?

File: graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
+import cats._
+import cats.data._
 import cats.mtl.implicits._
+
 import org.scalatest._
 
 class GraphzSpec extends FunSpec with Matchers with BeforeAndAfterEach with AppendedClues {
 
-  type Effect[A] = StateT[Id, StringBuffer, A]
+  type Effect[A] = StateT[Eval, StringBuffer, A]
```
Comment (ArturGajowy) [inline graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: Why not Coeval? Monix docs have:

```
at the moment of writing this, the Monix Coeval takes care of error handling, while the Cats Eval does not, providing operations for recovery, thus also working well as a replacement for Scala’s Try type.
```

How does Eval handle errors in the wrapped code? Will `.value` simply throw an exception?

**Review by ArturGajowy — APPROVED** [review]

File: graphz/src/main/scala/coop/rchain/graphz/Graphz.scala (hunk 1 above)
Comment (sebtomba) [reply graphz/src/main/scala/coop/rchain/graphz/Graphz.scala]: @ArturGajowy You are right.

File: graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala (hunk 2 above)
Comment (sebtomba) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: @ArturGajowy I see your point. But let's stick to `cats` for now. All of our current `GraphSerializer` instances are pure and won't throw any (non-fatal) exceptions.

Comment (marcin-rzeznicki) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: Irrelevant side note: `StateT[Eval, A, B]` is just `State[A, B]`

Comment (ArturGajowy) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: @marcin-rzeznicki are you sure that State[A, B] is stacksafe?

Comment (marcin-rzeznicki) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: Yeah: 
https://github.com/typelevel/cats/blob/8c031e5a4ed63a8efc34c196f95ccfcfea7eeb77/core/src/main/scala/cats/data/package.scala#L58

https://github.com/typelevel/cats/blob/ed99e170ee65dfe9d03995dd72b33ec90ae9f926/tests/src/test/scala/cats/tests/IndexedStateTSuite.scala#L40

Comment (marcin-rzeznicki) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: Had it used `Id` then obviously it wouldn't have been stack safe. It strongly relies on stack-safety of the underlying `F` ie.
https://github.com/typelevel/cats/blob/ed99e170ee65dfe9d03995dd72b33ec90ae9f926/core/src/main/scala/cats/data/IndexedStateT.scala#L25

Comment (ArturGajowy) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: Oh, cool! Wasn't aware of the Eval in there :)

File: graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala [hunk 3]
```diff
@@ -334,14 +336,21 @@ class GraphzSpec extends FunSpec with Matchers with BeforeAndAfterEach with Appe
...
     }
 
+    it("Huge graph") { // test for a stack overflow
+      val graph = for {
+        g <- Graphz[Effect]("G", DiGraph)
+        _ <- (1 to 1000).toList.traverse(i => g.edge(s"e$i" -> s"e${i + 1}"))
+        _ <- g.close
+      } yield g
+      graph.show // ignore
+    }
```
Comment (marcin-rzeznicki) [inline graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala:346]: :+1:

File: node/src/main/scala/coop/rchain/node/api/DeployGrpcService.scala [hunk 4]
```diff
@@ -57,10 +57,10 @@ object DeployGrpcService {
         defer(BlockAPI.getBlock[F](q))
 
       override def visualizeDag(q: VisualizeDagQuery): Observable[GrpcEither] = {
-        type Effect[A] = StateT[Id, Vector[String], A]
+        type Effect[A] = StateT[Eval, Vector[String], A]
```
Comment (marcin-rzeznicki) [inline node/src/main/scala/coop/rchain/node/api/DeployGrpcService.scala]: Right - it's the same as `type Effect[A] = State[Vector[String], A]` so maybe you could simplify it

File: graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala (hunk 2 above)
Comment (marcin-rzeznicki) [reply graphz/src/test/scala/coop/rchain/graphz/GraphzSpec.scala]: Maybe simplify the type to `State[Sth, A]`

File: graphz/src/main/scala/coop/rchain/graphz/Graphz.scala [hunk 5]
```diff
@@ -2,22 +2,21 @@ package coop.rchain.graphz
...
-class StringSerializer[F[_]: Applicative: MonadState[?[_], StringBuffer]]
-    extends GraphSerializer[F] {
+class StringSerializer[F[_]: MonadState[?[_], StringBuffer]] extends GraphSerializer[F] {
   def push(str: String, suffix: String): F[Unit] =
     MonadState[F, StringBuffer].modify(sb => sb.append(str + suffix))
 }
 
-class ListSerializer[F[_]: Applicative: MonadState[?[_], Vector[String]]]
-    extends GraphSerializer[F] {
+class ListSerializer[F[_]: MonadState[?[_], Vector[String]]] extends GraphSerializer[F] {
```
Comment (marcin-rzeznicki) [inline graphz/src/main/scala/coop/rchain/graphz/Graphz.scala:19]: Why do we actually need it here? :thinking: And why does a serializer return `Unit`? :thinking:

**Review by marcin-rzeznicki — APPROVED** [review]

**Pull request merged**